src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx

```diff
@@ -0,0 +1,237 @@
+
+import './TableColumnSetup.scss';
+
+const b = block('table-column-setup');
+
+type Item = TableColumnSetupItem;
+
+interface TableColumnSetupProps {
+    // for Button
+    disabled?: boolean;
+    switcher?: React.ReactElement | undefined;
+
+    // for List
+    items: Item[];
+    sortable?: boolean;
+    filterable?: boolean;
+
+    onUpdate: (updated: Item[]) => void;
+    popupWidth?: string;
+    popupPlacement?: PopperPlacement;
+    getItemTitle?: (item: Item) => TableColumnSetupItem['title'];
+    showStatus?: boolean;
+    className?: string;
+}
+
+export const TableColumnSetup = (props: TableColumnSetupProps) => {
+    const {
+        switcher,
+        disabled,
+        popupWidth,
+        popupPlacement,
+        className,
+        items: propsitems,
+        getItemTitle = (item: Item) => item.title,
+        sortable = true,
+        filterable = false,
+        showStatus,
+    } = props;
+
+    const [focused, setFocused] = React.useState(false);
+    const [items, setItems] = React.useState<Item[]>([]);
+    const [currentItems, setCurrentItems] = React.useState<Item[]>([]);
+    const [requiredItems, setRequiredItems] = React.useState<Item[]>([]);
+
+    const refControl = React.useRef(null);
+
+    const LIST_ITEM_HEIGHT = 36;
+
+    const getRequiredItems = (list: Item[]) =>
+        list
+            .filter(({required}) => required)
+            .map((column) => ({
+                ...column,
+                disabled: true,
+            }));
+
+    const getConfigurableItems = (list: Item[]) => list.filter(({required}) => !required);
+
+    React.useEffect(() => {
+        if (propsitems !== items) {
+            setItems(propsitems);
+            setRequiredItems(getRequiredItems(propsitems));
+            setCurrentItems(getConfigurableItems(propsitems));
+        }
+    }, [items, propsitems]);
+
+    const getListHeight = (list: Item[]) => {
+        const itemHeight = LIST_ITEM_HEIGHT;
+
+        return Math.min(5, list.length) * itemHeight + itemHeight / 2;
+    };
+
+    const getRequiredListHeight = (list: Item[]) => {
+        return list.length * LIST_ITEM_HEIGHT;
+    };
+
+    const onUpdate = (value: Item[]) => {
+        setCurrentItems(value);
+    };
+
+    const setInitialState = () => {
+        setFocused(false);
+        setRequiredItems(getRequiredItems(items));
+        setCurrentItems(getConfigurableItems(items));
+    };
+
+    const onClosePopup = () => {
+        setInitialState();
+    };
+
+    const onControlClick = () => {
+        if (!disabled) {
+            setFocused(!focused);
+            setRequiredItems(getRequiredItems(items));
+            setCurrentItems(getConfigurableItems(items));
+        }
+    };
+
+    const onApplyClick = () => {
+        setInitialState();
+
+        const newItems = requiredItems.concat(currentItems);
+
+        if (items !== newItems) {
+            props.onUpdate(newItems);
+        }
+    };
+
+    const renderItem = (item: Item) => {
```

Comment:
Let's organize functions here:
- `on* ` methods rename to `handle*` and put them together
- `render*` methods put together and place them after `handle*` methods and in the very end, right before return
- other methods place at the beginning